### ISSUE TITLE
Support for SvelteKit

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno#17248]
Prompted by discussion on Discord I tried to get SvelteKit working in Deno.

After an hour of debugging I found the main blocker: by default SvelteKit installed Node adapter, which does a bit of setup that is not needed in Deno and makes the project crash, namely:
https://github.com/sveltejs/kit/blob/e7bc0be2b25aff5ac151e3d83b771ad80cac1ab8/packages/kit/src/exports/node/polyfills.js#L6-L28
which installs polyfills from `undici` for various Web APIs like `fetch` or `Response`. These objects work slightly differently in Deno and arguably installing them in Deno is not needed. 

I think the way forward would be to get a Deno dedicated adapter that would essentially be a noop as there's no need to change anything to make SvelteKit work.

Of course it would be very nice if `deno run -A npm:create-svelte@latest` would set up a project in a way that requires no changes (namely making specifiers use `npm:` prefixes).

**Comment by benmccann** [comment]
Hi @bartlomieju, thanks for taking a look to support SvelteKit! I work on SvelteKit and wanted to better understand the issue to know if there's anything we're doing wrong on our side. Newer versions of Node provide those APIs out-of-the-box, but older versions don't. SvelteKit polyfills those APIs only when they're not already available.

> These objects work slightly differently in Deno and arguably installing them in Deno is not needed.

I know very little about Deno. I'd assume if Deno has those APIs then SvelteKit would not polyfill them, but if it doesn't have those APIs then SvelteKit will probably break if they don't get polyfilled. Does Deno do something differently than Node here such as not adding the APIs to `globalThis`? How should users utilize the APIs in this case?

> I think the way forward would be to get a Deno dedicated adapter that would essentially be a noop as there's no need to change anything to make SvelteKit work.

To clarify, would that adapter live in SvelteKit?

>Of course it would be very nice if deno run -A npm:create-svelte@latest would set up a project in a way that requires no changes (namely making specifiers use npm: prefixes).

Assuming this is something we'd have to implement on our side, can you clarify what "making specifiers use npm: prefixes" means?

**Comment by bartlomieju** [comment]
Hi @benmccann, thanks for taking the time.

> Newer versions of Node provide those APIs out-of-the-box, but older versions don't. SvelteKit polyfills those APIs only when they're not already available.

I haven't seen this bit - from what I saw in the Node adapter these polyfills were installed unconditionally. I assume there must be some check for Node version that causes the polyfills to be installed. Which version would be required? When polyfilling Node APIs we return version `v18.12.1` (latest LTS Node.js version).

> I know very little about Deno. I'd assume if Deno has those APIs then SvelteKit would not polyfill them, but if it doesn't have those APIs then SvelteKit will probably break if they don't get polyfilled. Does Deno do something differently than Node here such as not adding the APIs to globalThis? How should users utilize the APIs in this case?

They work the same in Deno as in browsers or with `undici`, they are present on `globalThis`. They work the same for intents and purposes for users, but internally they are implemented differently than in browsers or `undici` and some of our APIs break when certain non-public fields are not present. This becomes non-issue if the polyfills are not installed.

> To clarify, would that adapter live in SvelteKit?

I think that's a question for you, whether you're interested in providing an official adapter for Deno, I'll be glad to help in any capacity to make it happen :)

> Assuming this is something we'd have to implement on our side, can you clarify what "making specifiers use npm: prefixes" means?

Deno has now built-in support for npm - we do that by using `npm:` specifiers - more info on that here: https://deno.land/manual@v1.29.1/node/npm_specifiers. Long story, short: we currently don't rely (or actually use at all) `package.json` file to figure out dependencies needed, but instead look at all imports with `npm:` specifiers to pull relevant packages (and then recursively analyzing them). What I meant by that comment was that it would be nice if Svelte's generator would for example put `import svelte from "npm:svelte"` instead of `import svelte from "svelte"` when it sees it's being run in Deno (eg. by checking for `typeof globalThis.Deno !== "undefined"`. Though I think it's a secondary issue at this point - we already added custom templates to https://github.com/bluwy/create-vite-extra that take care of that and we're debating internally if we should support `package.json` in user code.

Thanks for dropping by and helping with this problem!

**Comment by benmccann** [comment]
> these polyfills were installed unconditionally

Oops! My bad. That was outdated info. It seems as of https://github.com/sveltejs/kit/pull/7675 that you're correct. It appears we had to do it unconditionally to overwrite the Node-provided version which is an older version of Undici and is broken (https://github.com/sveltejs/kit/issues/7673). Maybe with Node 20 or SvelteKit 2 we'll be able to do something nicer :smile: 

>  from what I saw in the Node adapter

Ah! I was pointed here by [an issue in the SvelteKit repo](https://github.com/sveltejs/kit/issues/8324) which talked about the `dev` command, which I guess was broken by the fact that we now polyfill unconditionally. The adapters are only run with the `build` command. The node adapter was really only meant for Node.js, but we do have a Deno adapter that a community member developed and I'm assuming this issue would not be present there: https://github.com/pluvial/svelte-adapter-deno.

I'm not quite sure what the best fix for the dev mode issue would be. Maybe we could detect node vs deno and only install the polyfills on Node. I'd prefer if there were a nicer way to do this, but maybe it's just temporary until Node 20 / SvelteKit 2 anyway so we can live with it being imperfect. I'll have to check with the other maintainers though

**Comment by bartlomieju** [comment]
> Oops! My bad. That was outdated info. It seems as of https://github.com/sveltejs/kit/pull/7675 that you're correct. It appears we had to do it unconditionally to overwrite the Node-provided version which is an older version of Undici and is broken (https://github.com/sveltejs/kit/issues/7673). Maybe with Node 20 or SvelteKit 2 we'll be able to do something nicer 😄

Thanks for a quick response. That's exactly the problem that I've faced.

> I'm not quite sure what the best fix for the dev mode issue would be. Maybe we could detect node vs deno and only install the polyfills on Node. I'd prefer if there were a nicer way to do this, but maybe it's just temporary until Node 20 / SvelteKit 2 anyway so we can live with it being imperfect. I'll have to check with the other maintainers though

A bit quick and dirty, but that would solve the problem at hand and would allow us and Deno users to leverage Svelte with Vite. I'm eager to hear what you decide to do. Let me know if there's anything I can help with to move this issue forward.

**Comment by benmccann** [comment]
Hopefully this is fixed in SvelteKit 1.0.4: https://github.com/sveltejs/kit/blob/master/packages/kit/CHANGELOG.md#104

Thanks for tracking down the issue and sharing enough details to help get it fixed! Let me know if there's anything else to be done on the SvelteKit side.

**Comment by bartlomieju** [comment]
> Hopefully this is fixed in SvelteKit 1.0.4: https://github.com/sveltejs/kit/blob/master/packages/kit/CHANGELOG.md#104
> 
> Thanks for tracking down the issue and sharing enough details to help get it fixed! Let me know if there's anything else to be done on the SvelteKit side.

Thank for a quick turnaround @benmccann, I can confirm that latest version fixes the problem. I'm gonna keep this issue open for a little longer as we might take approach of supporting SvelteKit running in Vite as-is, without any changes (eg. `npm:` specifiers).

**Comment by jpaquim** [comment]
I also confirm that updating fixes the issue with running the development server. I've also been trying to integrate svelte-adapter-deno to generate the production build itself for Deno: https://github.com/jpaquim/sveltekit-deno

I'm getting this error, though:
```
error: Uncaught TypeError: Could not resolve 'file:///Users/<user>/sveltekit-deno/.svelte-kit/output/server/index.js' from 'file:///Users/<user>/sveltekit-deno/node_modules/.deno/@sveltejs+kit@1.0.5/node_modules/@sveltejs/kit/src/core/prerender/prerender.js'.
  Caused by:
    could not find npm package for 'file:///Users/<user>/sveltekit-deno/.svelte-kit/output/server/index.js'
    at async prerender (file:///Users/<user>/sveltekit-deno/node_modules/.deno/@sveltejs+kit@1.0.5/node_modules/@sveltejs/kit/src/core/prerender/prerender.js:100:31)
[vite-plugin-sveltekit-compile] Prerendering failed with code 1
error during build:
Error: Prerendering failed with code 1
    at ChildProcess.<anonymous> (file:///Users/<user>/sveltekit-deno/node_modules/.deno/@sveltejs+kit@1.0.5/node_modules/@sveltejs/kit/src/exports/vite/index.js:551:15)
    at ChildProcess.emit (https://deno.land/std@0.171.0/node/_events.mjs:379:28)
    at https://deno.land/std@0.171.0/node/internal/child_process.ts:229:16
```
The file (`.svelte-kit/output/server/index.js`) is generated correctly, but for some reason importing it seems to be failing, maybe due to the way the dynamic import is being built here? https://github.com/sveltejs/kit/blob/master/packages/kit/src/core/prerender/prerender.js#L100

**Comment by jpaquim** [comment]
I also had to manually remove the `@neoconfetti/svelte` import as it wasn't working even with the `npm:` specifier. Same thing with the `@import '@fontsource/fira-mono'` in `styles.css`, commented out for now.

**Comment by jakubdonovan** [comment]
> > Hopefully this is fixed in SvelteKit 1.0.4: https://github.com/sveltejs/kit/blob/master/packages/kit/CHANGELOG.md#104
> > Thanks for tracking down the issue and sharing enough details to help get it fixed! Let me know if there's anything else to be done on the SvelteKit side.
> 
> Thank for a quick turnaround @benmccann, I can confirm that latest version fixes the problem. I'm gonna keep this issue open for a little longer as we might take approach of supporting SvelteKit running in Vite as-is, without any changes (eg. `npm:` specifiers).

It looks like the sk patch was reviewed and merged by rich harris yesterday. https://github.com/sveltejs/kit/pull/8338

What's the process of installing sveltekit in a deno project now? I've ran `deno run -A npm:create-vite-extra` and selected the `deno-svelte` template with `typescript` and then went to `vite.config.mts` and replaced the svelte plugin with sveltekit like so 
```ts
import { defineConfig } from 'npm:vite@^4.0.0'
import { svelte } from 'npm:@sveltejs/vite-plugin-svelte@^2.0.0'
import { sveltekit } from 'npm:@sveltejs/kit/vite'
import 'npm:svelte@^3.54.0'

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [sveltekit()]
})
``` 
ran `deno task dev` and still got the pesky undefined body error. When looking through my `deno.lock`, I noticed that the pre-patched version of sveltekit was imported and that my file structure has things missing (app.html, the routes folder, no +page.svelte) etc

**Comment by bartlomieju** [comment]
> What's the process of installing sveltekit in a deno project now? I've ran deno run -A npm:create-vite-extra and selected the deno-svelte template with typescript and then went to vite.config.mts and replaced the svelte plugin with sveltekit like so

Most likely the template needs to be updated - I created these templates about 3 months ago so they might be outdated by now.

> ran deno task dev and still got the pesky undefined body error. When looking through my deno.lock, I noticed that the pre-patched version of sveltekit (1.0.1) was imported and that my file structure has things missing (app.html, the routes folder, no +page.svelte) etc

If you have used Svelte before and used unversioned import you might get a version from cache. Either use version import or run with `--reload` to force Deno to pull latest version.

> A clarification will be highly appreciated! How would we go about getting a deno-sveltekit template in create-vite-extra?

I just opened a PR to `create-vite-extra` that essentially copy-pasted existing templates and adjusted them to be Deno compatible (by changing imports to use `npm:` specifiers and list all dependencies in `vite.config.mts`.

**Comment by fernandolguevara** [comment]
I'm trying to run 
`deno run -A --unstable --node-modules-dir npm:vite build`
on a [deno + sveltekit + unocss project](https://github.com/fernandolguevara/deno-sk-uno/)

but got this error

```sh
error: Uncaught TypeError: Could not resolve 'file:///Users/anon/deno-sk-uno/.svelte-kit/output/server/index.js' from 'file:///Users/anon/deno-sk-uno/node_modules/.deno/@sveltejs+kit@1.0.10/node_modules/@sveltejs/kit/src/core/prerender/prerender.js'.
  Caused by:
    could not find npm package for 'file:///Users/anon/deno-sk-uno/.svelte-kit/output/server/index.js'
    at async prerender (file:///Users/anon/deno-sk-uno/node_modules/.deno/@sveltejs+kit@1.0.10/node_modules/@sveltejs/kit/src/core/prerender/prerender.js:102:31)
[vite-plugin-sveltekit-compile] Prerendering failed with code 1
error during build:
Error: Prerendering failed with code 1
    at ChildProcess.<anonymous> (file:///Users/anon/deno-sk-uno/node_modules/.deno/@sveltejs+kit@1.0.10/node_modules/@sveltejs/kit/src/exports/vite/index.js:551:15)
    at ChildProcess.emit (https://deno.land/std@0.173.0/node/_events.mjs:379:28)
    at https://deno.land/std@0.173.0/node/internal/child_process.ts:229:16
```

The issue has occurred on the child process spawned for the pre-render stage...
the generated files on `output/*.js` don't follow deno import syntaxt for `npm` packages

**Comment by benmccann** [comment]
It appears this will be fixed by future updates to Deno: https://github.com/denoland/deno/issues/17475

**Comment by bartlomieju** [comment]
> running deno task build in a skeleton-app created with create-svelte results in an error 

```
develop/reproductions/denokit ❯ deno task build
Warning Currently only basic package.json `scripts` are supported. Programs like `rimraf` or `cross-env` will not work correctly. This will be fixed in the upcoming release.
Task build vite build

vite v4.1.4 building SSR bundle for production...
transforming (72) src/routes/styles.css"confetti" is imported from external module "@neoconfetti/svelte" but never used in "src/routes/sverdle/+page.svelte".
✓ 74 modules transformed.
6:23:26 PM [vite-plugin-svelte] ssr compile done.
package    files      time       avg
denokit        9    91.3ms    10.1ms
error: Uncaught Error: Argument list too long (os error 7)
    at spawnChildInner (internal:runtime/js/40_spawn.js:45:17)
    at spawnChild (internal:runtime/js/40_spawn.js:65:10)
    at Command.spawn (internal:runtime/js/40_spawn.js:322:12)
    at new ChildProcess (internal:deno_node/polyfills/internal/child_process.ts:108:16)
    at spawn (internal:deno_node/polyfills/child_process.ts:102:12)
    at Object.fork (internal:deno_node/polyfills/child_process.ts:93:12)
    at file:///home/dominikg/develop/reproductions/denokit/node_modules/.deno/@sveltejs+kit@1.8.3/node_modules/@sveltejs/kit/src/utils/fork.js:39:32
    at new Promise (<anonymous>)
    at fn (file:///home/dominikg/develop/reproductions/denokit/node_modules/.deno/@sveltejs+kit@1.8.3/node_modules/@sveltejs/kit/src/utils/fork.js:38:10)
    at Object.handler (file:///home/dominikg/develop/reproductions/denokit/node_modules/.deno/@sveltejs+kit@1.8.3/node_modules/@sveltejs/kit/src/exports/vite/index.js:648:28)
```

Reported on Discord by @dominikg

**Comment by fernandolguevara** [comment]
@bartlomieju It is not possible to run the `vite build` on `deno` for a `svelte-kit project` ... it seems `deno`  doesn't implement `process.send`  used in the sveltekit prerender stage for the ipc

https://github.com/sveltejs/kit/blob/master/packages/kit/src/core/postbuild/prerender.js#L16 
https://github.com/sveltejs/kit/blob/master/packages/kit/src/utils/fork.js#L14

**Comment by benmccann** [comment]
The reason SvelteKit does prerendering in a subprocess is so that we can call `process.exit` to ensure it's fully killed (e.g. to support cases where the user's app does something like leave a database connection we need to kill) without killing the main process (see https://github.com/sveltejs/kit/issues/5306).

I'm not familiar with web workers, but we would need to be able to call `process.exit` from one without killing the main process. It sounds like that's what happens so it should be a viable alternative: https://nodejs.org/api/worker_threads.html#class-worker

If anyone wants to send a PR, this sounds like something we could accept a PR for in SvelteKit assuming it doesn't massively complicate things.

**Comment by andocodes** [comment]
Could anyone gauge how much effort it would take for say, Svelte and SvelteKit to provide an actual native Deno package? Seems like a huge opportunity?

**Comment by EthanThatOneKid** [comment]
> Could anyone gauge how much effort it would take for say, Svelte and SvelteKit to provide an actual native Deno package? Seems like a huge opportunity?

As a community member, it's challenging to gauge the effort it would take to provide a native Deno package for Svelte and SvelteKit, and it's better left to SvelteKit or Deno core developers. However, there is some relevant information to consider.

Currently, we can use the Svelte compiler module from NPM in Deno, as shown in the example below:

```ts
import { compile } from "npm:svelte/compiler";

const result = compile("Hello world", {
    format: "esm",
    filename: "hello-world.svelte",
});

console.log({ result });
```

This allows us to use the Svelte compiler in Deno. However, SvelteKit, the framework, relies on other Svelte language tools that are not fully compatible with Deno, such as the Svelte LSP which is built for the Node.js toolchain.

It appears that several Svelte tools would need to be refactored in order to fully support Deno. Therefore, determining the effort required to provide a native Deno package for Svelte and SvelteKit would require closer collaboration between the SvelteKit and Deno core developers.

**Comment by benmccann** [comment]
> SvelteKit, the framework, relies on other Svelte language tools that are not fully compatible with Deno, such as the Svelte LSP which is built for the Node.js toolchain

I guess you're talking about the `check` script that's included in new projects when you choose to enable type checking in the new project creation wizard? While `svelte-check` is a very nice tool for type checking your `.svelte` files, you can still fully use SvelteKit without it. It's probably best to discuss `svelte-check` in a separate ticket because it's a different project with a different set of developers from SvelteKit. Unfortunately I don't know anything about the `svelte-check` internals.

For SvelteKit itself, the only thing I'm aware of that's needed is switching the subprocess to a worker as mentioned in https://github.com/denoland/deno/issues/17248#issuecomment-1499382376. It's not a priority for the Svelte team as we focus on Svelte 4, but if anyone wants to send a PR I'm happy to advise and we will review it!

**Comment by slacktracer** [comment]
I very much hope this moves forward, and soon. By the issues I've been looking Deno and SvelteKit don't know each other very well but having spent a good chunk of time getting to know both I am very confident they would be a match made in heaven... 🤞🏼

**Comment by Afrowave** [comment]
This is an excellent thread. I recently committed all my future frontend work to be done in SvelteKit, having come from a more "traditional" Design > HTML5/CSS > frontend route.  
  
SvelteKit works the way I think frontend frameworks should. 😁 Deno is the future for "tidy" front-end web development. Once we throw in WebAssembly and Domain-Specific-Languages like Rust and Python, wrapped in single binary executable, then we will be in business. 😃

**Comment by dominikg** [comment]
fyi, you can already use deno to run sveltekit applications, it's building that doesn't work currently. 
So eg. have a multi-stage Dockerfile where the first stage builds your app with node, then copies the result to a runner stage from a deno image of the same arch, call deno with appropriate permission flags and you're good to go.

**Comment by Afrowave** [comment]
Thank you @dominikg,  
  
That is very good to know. I should try this out and report back. Better still, do you have or where can one find a repo with something like you have explained?

**Comment by Afrowave** [comment]
Thank you @dominikg,  
  
That is very good to know. I should try this out and report back. Better still, do you have or where can one find a repo with something like you have explained?

**Comment by about-hector** [comment]
> fyi, you can already use deno to run sveltekit applications, it's building that doesn't work currently. So eg. have a multi-stage Dockerfile where the first stage builds your app with node, then copies the result to a runner stage from a deno image of the same arch, call deno with appropriate permission flags and you're good to go.

@dominikg do you mind sharing an example dockerfile for the basic svelte starter app?
I'm not quite getting the part where you copy the result to the runner stage. Most likely my bad for not understanding deno and docker well enough.

**Comment by bartlomieju** [comment]
@benmccann thanks for the ping. I'm OOO this week. I will verify that everything works as expected and that Deno is compatible with Svelte v4 (congratulations!) over the weekend and close the issue then.

**Comment by endigma** [comment]
If this is to be closed as complete, is there a full example of how to get SvelteKit working with deno? Bits and pieces aren't great for future reference.

**Comment by bartlomieju** [comment]
@benmccann sorry for slow response. I just tried svelte@3.54 and @sveltejs/kit@1.5.0 and while `deno task dev` works fine I get this error for `deno task build`:
```
error during build:
Error: Unexpected early exit. This happens when Promises returned by plugins cannot resolve. Unfinished hook action(s) on exit:
(vite-plugin-sveltekit-compile) writeBundle
    at process.handleBeforeExit (file:///Users/ib/dev/try_svelte_kit/my-app/node_modules/.deno/vite@4.3.9/node_modules/rollup/dist/es/shared/node-entry.js:25670:28)
    at Object.onceWrapper (ext:deno_node/_events.mjs:507:26)
    at process.emit (ext:deno_node/_events.mjs:382:28)
    at process.emit (ext:deno_node/process.ts:343:26)
    at ext:deno_node/process.ts:535:17
    at innerInvokeEventListeners (ext:deno_web/02_event.js:792:7)
    at invokeEventListeners (ext:deno_web/02_event.js:839:5)
    at dispatch (ext:deno_web/02_event.js:696:9)
    at dispatchEvent (ext:deno_web/02_event.js:1101:12)
    at [ext:cli/worker.rs:161:38]:1:1

```

Trying `create-svelte@5.0.1` (`deno run -A --unstable --reload npm:create-svelte@latest my-app`) I get:
```

create-svelte version 5.0.1

┌  Welcome to SvelteKit!
error: Uncaught Error: Not implemented: net.Socket.prototype.constructor with fd option
    at notImplemented (ext:deno_node/_utils.ts:7:11)
    at new Socket (ext:deno_node/net.ts:453:13)
    at new WriteStream (ext:deno_node/tty.ts:18:8)
    at ED.prompt (file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/@clack/core/0.3.2/dist/index.mjs:9:693)
    at Module.ee (file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/@clack/prompts/0.6.3/dist/index.mjs:28:7)
    at template (file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/create-svelte/5.0.1/bin.js:48:6)
    at Module.$e (file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/@clack/prompts/0.6.3/dist/index.mjs:80:345)
    at file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/create-svelte/5.0.1/bin.js:45:25
```

So it seems we got a few more bugs to iron out. I'll try to prioritize this work in July.

**Comment by GTime** [comment]
> @benmccann sorry for slow response. I just tried svelte@3.54 and @sveltejs/kit@1.5.0 and while `deno task dev` works fine I get this error for `deno task build`:
> 
> ```
> error during build:
> Error: Unexpected early exit. This happens when Promises returned by plugins cannot resolve. Unfinished hook action(s) on exit:
> (vite-plugin-sveltekit-compile) writeBundle
>     at process.handleBeforeExit (file:///Users/ib/dev/try_svelte_kit/my-app/node_modules/.deno/vite@4.3.9/node_modules/rollup/dist/es/shared/node-entry.js:25670:28)
>     at Object.onceWrapper (ext:deno_node/_events.mjs:507:26)
>     at process.emit (ext:deno_node/_events.mjs:382:28)
>     at process.emit (ext:deno_node/process.ts:343:26)
>     at ext:deno_node/process.ts:535:17
>     at innerInvokeEventListeners (ext:deno_web/02_event.js:792:7)
>     at invokeEventListeners (ext:deno_web/02_event.js:839:5)
>     at dispatch (ext:deno_web/02_event.js:696:9)
>     at dispatchEvent (ext:deno_web/02_event.js:1101:12)
>     at [ext:cli/worker.rs:161:38]:1:1
> ```
> 
> Trying `create-svelte@5.0.1` (`deno run -A --unstable --reload npm:create-svelte@latest my-app`) I get:
> 
> ```
> 
> create-svelte version 5.0.1
> 
> ┌  Welcome to SvelteKit!
> error: Uncaught Error: Not implemented: net.Socket.prototype.constructor with fd option
>     at notImplemented (ext:deno_node/_utils.ts:7:11)
>     at new Socket (ext:deno_node/net.ts:453:13)
>     at new WriteStream (ext:deno_node/tty.ts:18:8)
>     at ED.prompt (file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/@clack/core/0.3.2/dist/index.mjs:9:693)
>     at Module.ee (file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/@clack/prompts/0.6.3/dist/index.mjs:28:7)
>     at template (file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/create-svelte/5.0.1/bin.js:48:6)
>     at Module.$e (file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/@clack/prompts/0.6.3/dist/index.mjs:80:345)
>     at file:///Users/ib/Library/Caches/deno/npm/registry.npmjs.org/create-svelte/5.0.1/bin.js:45:25
> ```
> 
> So it seems we got a few more bugs to iron out. I'll try to prioritize this work in July.

I got same issue:

```sh
siki@siki ~/Lab> deno run -A --unstable --reload npm:create-svelte@latest sveltekit-deno

create-svelte version 5.0.2

┌  Welcome to SvelteKit!
error: Uncaught Error: Not implemented: net.Socket.prototype.constructor with fd option
    at notImplemented (ext:deno_node/_utils.ts:7:11)
    at new Socket (ext:deno_node/net.ts:453:13)
    at new WriteStream (ext:deno_node/tty.ts:18:8)
    at ED.prompt (file:///home/siki/.cache/deno/npm/registry.npmjs.org/@clack/core/0.3.2/dist/index.mjs:9:693)
    at Module.ee (file:///home/siki/.cache/deno/npm/registry.npmjs.org/@clack/prompts/0.6.3/dist/index.mjs:28:7)
    at template (file:///home/siki/.cache/deno/npm/registry.npmjs.org/create-svelte/5.0.2/bin.js:48:6)
    at Module.$e (file:///home/siki/.cache/deno/npm/registry.npmjs.org/@clack/prompts/0.6.3/dist/index.mjs:80:345)
    at file:///home/siki/.cache/deno/npm/registry.npmjs.org/create-svelte/5.0.2/bin.js:45:25
```

**Comment by fernandolguevara** [comment]
This issue is related to the tty implementation on deno (not implemented), the running script ask the runtime to read the user input for project settings, As a workaroud you can created the sveltekit project with npm/pnpm (node), 
then you can run the project with deno run the following command on your project folder 

`deno run -A --node-modules-dir npm:vite`

You can add this command on your package scripts or crate a deno.json

**Comment by vfrunza** [comment]
Following what @fernandolguevara suggested appears to work with updated svelte/kit packages (and with svelte 4).

**Comment by fernandolguevara** [comment]
@vfrunza If you want to publish your project on Deno Deploy, I'm using this adapter  [sveltekit-adapter-deno](https://github.com/pluvial/svelte-adapter-deno)

**Comment by fernandolguevara** [comment]
@happysalada 
did u try with this, from [vite-site](https://vitejs.dev/config/#using-environment-variables-in-config)
```ts
import { defineConfig, loadEnv } from 'vite'

export default defineConfig(({ command, mode }) => {
  // Load env file based on `mode` in the current working directory.
  // Set the third parameter to '' to load all env regardless of the `VITE_` prefix.
  const env = loadEnv(mode, process.cwd(), '') // here Deno.env
  return {
    // vite config
    define: {
      __APP_ENV__: JSON.stringify(env.APP_ENV),
    },
  }
})
```

**Comment by happysalada** [comment]
I didnt try no, i just saw the issue left open and assumed it was correct. Thank you for the suggestion!

**Comment by drewbitt** [comment]
So from reading the above, the blocker is some tty methods for create-svelte, but otherwise it works fine?

**Comment by begoon** [comment]
I finally found the sveltekit apaptor for deno deploy by @dbushell, which works for the current versions (at the time of writing this comment) of svelte/sveltekit and deno/deploy (including KV). I managed to build and run locally with deno, and then deploy to Deno Deploy.

```
pnpm i
Lockfile is up to date, resolution step is skipped

devDependencies:
+ @sveltejs/kit 1.24.1
+ svelte 4.2.0
+ svelte-adapter-deno 0.9.0
+ svelte-check 3.5.1
+ tslib 2.6.2
+ typescript 5.2.2
+ vite 4.4.9
```

https://github.com/dbushell/sveltekit-adapter-deno

https://dbushell.com/2023/06/26/sveltekit-oauth-deno-deploy/

**Comment by benmccann** [comment]
Not being able to run `create-svelte` (i.e. `npm create svelte@latest`) is blocked by https://github.com/denoland/deno/issues/6529 as described in https://github.com/denoland/deno/issues/6529#issuecomment-1432690559

**Comment by benmccann** [comment]
`create-svelte` should hopefully be fixed in the next version of Deno: https://github.com/denoland/deno/pull/20892

**Comment by h7kanna** [comment]
I built deno from main branch and this gave the below error

```
h7kanna@Harshas-MBP ~ % ./Repos/crates/deno/target/release/deno run -A --unstable --reload npm:create-svelte@latest sveltekit-deno

create-svelte version 5.1.1

┌  Welcome to SvelteKit!
│
◆  Which Svelte app template?
│  ● Skeleton project (Barebones scaffolding for your new SvelteKit app)
│  ○ SvelteKit demo app
│  ○ Library project
└
error: Uncaught Error: The _read() method is not implemented
    at ReadStream.Readable._read (ext:deno_node/_stream.mjs:3101:13)
    at ReadStream.Readable.read (ext:deno_node/_stream.mjs:2997:16)
    at resume_ (ext:deno_node/_stream.mjs:3347:16)
    at Array.processTicksAndRejections (ext:deno_node/_next_tick.ts:33:15)
    at eventLoopTick (ext:core/01_core.js:180:29)
```

**Comment by littledivy** [comment]
Opened another issue for `create-svelte` - https://github.com/denoland/deno/issues/21024

**Comment by benmccann** [comment]
`create-svelte` was fixed: https://github.com/denoland/deno/pull/21026

**Comment by h7kanna** [comment]
@bartlomieju @littledivy 
I just ran the create-svelte again and was able to create the project.
But there seems to be another issue.

'vite dev' is successful, able to run the app.
But 'vite build' seems to be stuck without any error messages.

```sh
../Repos/crates/deno/target/release/deno run -A --unstable --reload npm:create-svelte@latest sveltekit-deno
create-svelte version 5.1.1

┌  Welcome to SvelteKit!
│
◇  Which Svelte app template?
│  SvelteKit demo app
│
◇  Add type checking with TypeScript?
│  Yes, using TypeScript syntax
│
◇  Select additional options (use arrow keys/space bar)
│  Add ESLint for code linting, Add Prettier for code formatting, Add Playwright for browser testing, Add Vitest for unit testing
│
└  Your project is ready!

✔ Typescript
  Inside Svelte components, use <script lang="ts">

✔ ESLint
  https://github.com/sveltejs/eslint-plugin-svelte

✔ Prettier
  https://prettier.io/docs/en/options.html
  https://github.com/sveltejs/prettier-plugin-svelte#options

✔ Playwright
  https://playwright.dev

✔ Vitest
  https://vitest.dev

Install community-maintained integrations:
  https://github.com/svelte-add/svelte-add

Next steps:
  1: cd sveltekit-deno
  2: npm install
  3: git init && git add -A && git commit -m "Initial commit" (optional)
  4: npm run dev -- --open

To close the dev server, hit Ctrl-C

Stuck? Visit us at https://svelte.dev/chat
```
Update the package.json

```typescript
"scripts": {
   "dev": "/Users/h7kanna/Repos/crates/deno/target/release/deno run -A --node-modules-dir npm:vite dev",
   "build": "/Users/h7kanna/Repos/crates/deno/target/release/deno run -A --node-modules-dir npm:vite build",
}
```


```sh
cd sveltekit-deno
h7kanna@Harshas-MBP sveltekit-deno % npm run build

> sveltekit-deno@0.0.1 build
> /Users/h7kanna/Repos/crates/deno/target/release/deno run -A --node-modules-dir npm:vite build


vite v4.5.0 building SSR bundle for production...
"confetti" is imported from external module "@neoconfetti/svelte" but never used in "src/routes/sverdle/+page.svelte".
✓ 113 modules transformed.

```

**Comment by rracariu** [comment]
Creating an application works with `Deno 1.40.4` and `SvelteKit 2.5`.

Running the dev server is possible also, but accessing the demo page would trigger an error related to `AsyncLocalStorage`.
See https://github.com/denoland/deno/issues/18127#issuecomment-1932743332

`deno task build` would get stuck, as previous comments mention.

**Comment by birkskyum** [comment]
@rracariu , accessing the demo page works now, since #22740 

Running `deno task build` gives me:

```sh
➜ deno task build
Task build vite build
vite v5.1.6 building SSR bundle for production...
"confetti" is imported from external module "@neoconfetti/svelte" but never used in "src/routes/sverdle/+page.svelte".
✓ 120 modules transformed.
error: Uncaught (in worker "") (in promise) Error: require() of ES Module /Users/admin/repos/deno-kitchensink/my-svelt-app/node_modules/.deno/@sveltejs+kit@2.5.4/node_modules/@sveltejs/kit/src/core/postbuild/analyse.js from /Users/admin/repos/deno-kitchensink/my-svelt-app not supported. Instead change the require to a dynamic import() which is available in all CommonJS modules.
    at createRequireEsmError (node:module:740:15)
    at Object.Module._extensions..js (node:module:729:13)
    at Module.load (node:module:643:32)
    at Function.Module._load (node:module:524:12)
    at Module.require (node:module:662:19)
    at require (node:module:776:16)
    at data:text/javascript,(async function() {c......d/analyse.js");})();:1:163
error: Uncaught (in promise) Error: Unhandled error. ([Object: null prototype] {
  message: 'Uncaught (in promise) Error: require() of ES Module /Users/admin/repos/deno-kitchensink/my-svelt-app/node_modules/.deno/@sveltejs+kit@2.5.4/node_modules/@sveltejs/kit/src/core/postbuild/analyse.js from /Users/admin/repos/deno-kitchensink/my-svelt-app not supported. Instead change the require to a dynamic import() which is available in all CommonJS modules.',
  fileName: 'node:module',
  lineNumber: 740,
  columnNumber: 15
})
    at NodeWorker.emit (ext:deno_node/_events.mjs:371:17)
    at NodeWorker.#handleError (node:worker_threads:155:10)
    at NodeWorker.#pollControl (node:worker_threads:175:30)
    at eventLoopTick (ext:core/01_core.js:169:7)
```

**Comment by rracariu** [comment]
> @rracariu , accessing the demo page works now, since #22740
> 
> Running `deno task build` on the other hand gives me:
> 
> ```shell
> ➜ deno task build
> Task build vite build
> vite v5.1.6 building SSR bundle for production...
> "confetti" is imported from external module "@neoconfetti/svelte" but never used in "src/routes/sverdle/+page.svelte".
> ✓ 120 modules transformed.
> error: Uncaught (in worker "") (in promise) Error: require() of ES Module /Users/admin/repos/deno-kitchensink/my-svelt-app/node_modules/.deno/@sveltejs+kit@2.5.4/node_modules/@sveltejs/kit/src/core/postbuild/analyse.js from /Users/admin/repos/deno-kitchensink/my-svelt-app not supported. Instead change the require to a dynamic import() which is available in all CommonJS modules.
>     at createRequireEsmError (node:module:740:15)
>     at Object.Module._extensions..js (node:module:729:13)
>     at Module.load (node:module:643:32)
>     at Function.Module._load (node:module:524:12)
>     at Module.require (node:module:662:19)
>     at require (node:module:776:16)
>     at data:text/javascript,(async function() {c......d/analyse.js");})();:1:163
> error: Uncaught (in promise) Error: Unhandled error. ([Object: null prototype] {
>   message: 'Uncaught (in promise) Error: require() of ES Module /Users/admin/repos/deno-kitchensink/my-svelt-app/node_modules/.deno/@sveltejs+kit@2.5.4/node_modules/@sveltejs/kit/src/core/postbuild/analyse.js from /Users/admin/repos/deno-kitchensink/my-svelt-app not supported. Instead change the require to a dynamic import() which is available in all CommonJS modules.',
>   fileName: 'node:module',
>   lineNumber: 740,
>   columnNumber: 15
> })
>     at NodeWorker.emit (ext:deno_node/_events.mjs:371:17)
>     at NodeWorker.#handleError (node:worker_threads:155:10)
>     at NodeWorker.#pollControl (node:worker_threads:175:30)
>     at eventLoopTick (ext:core/01_core.js:169:7)
> ```

Yes, I was able to run it in dev mode as well. Fixing the build command would make it finally work end-to-end.
ATM I'm using `npm run build` for the final build step, but already miles better DX as I can use all of Deno's API in the server side of the app I'm building and also use Deno deploy for the release.

**Comment by bartlomieju** [comment]
This problem will be fixed by https://github.com/denoland/deno/pull/22977 and https://github.com/denoland/deno/pull/22841

**Comment by Rudo2204** [comment]
I got `deno task dev` working but `deno task build` is giving me:
```
$ deno task build
Task build vite build
vite v5.2.6 building SSR bundle for production...
"confetti" is imported from external module "@neoconfetti/svelte" but never used in "src/routes/sverdle/+page.svelte".
✓ 120 modules transformed.
x Build failed in 423ms
error during build:
Error: Unexpected early exit. This happens when Promises returned by plugins cannot resolve. Unfinished hook action(s) on exit:
(vite-plugin-sveltekit-compile) writeBundle
    at process.handleBeforeExit (file:///home/rudo/code/my-app/node_modules/.deno/vite@5.2.6/node_modules/rollup/dist/es/shared/node-entry.js:19932:28)
    at Object.onceWrapper (ext:deno_node/_events.mjs:511:26)
    at process.emit (ext:deno_node/_events.mjs:386:28)
    at process.emit (node:process:378:20)
    at processOnBeforeUnload (node:process:572:11)
    at innerInvokeEventListeners (ext:deno_web/02_event.js:754:7)
    at invokeEventListeners (ext:deno_web/02_event.js:801:5)
    at dispatch (ext:deno_web/02_event.js:658:9)
    at dispatchEvent (ext:deno_web/02_event.js:1043:12)
    at [ext:cli/worker.rs:224:38]:1:1
```

I'm using `vite/5.2.6 linux-x64 node-v20.11.1` and `node v20.12.0`

**Comment by lfmaciasr** [comment]
After upgrading to Deno v1.42.0, which fixes https://github.com/denoland/deno/pull/22977 and https://github.com/denoland/deno/pull/22841 I get the following error on Windows:

deno run -A --unstable --reload npm:create-svelte@latest sveltekit-deno
⚠️  The `--unstable` flag is deprecated and will be removed in Deno 2.0. Use granular `--unstable-*` flags instead.
Learn more at: https://docs.deno.com/runtime/manual/tools/unstable_flags

create-svelte version 6.0.10

T  Welcome to SvelteKit!
|
*  Which Svelte app template?
|  > SvelteKit demo app (A demo app showcasing some of the features of SvelteKit - play a word guessing game that works without
JavaScript!)
|    Skeleton project
|    Library project
—
error: Top-level await promise never resolved
    at <anonymous> (file:///C:/Users/lfmaciasr/AppData/Local/deno/npm/registry.npmjs.org/create-svelte/6.0.10/bin.js:45:17)

**Comment by birkskyum** [comment]
Try setting DENO_FUTURR=1 and the right granular unstable flags in a demo.json

**Comment by lfmaciasr** [comment]
No luck here.
Maybe the error
error: Top-level await promise never resolved
at (file:///C:/Users/lfmaciasr/AppData/Local/deno/npm/registry.npmjs.org/create-svelte/6.0.10/bin.js:45:17)
is related to some other change?

**Comment by Rudo2204** [comment]
> After upgrading to Deno v1.42.0, which fixes #22977 and #22841, I get the following error on Windows:
> 
> deno run -A --unstable --reload npm:create-svelte@latest sveltekit-deno ⚠️ The `--unstable` flag is deprecated and will be removed in Deno 2.0. Use granular `--unstable-*` flags instead. Learn more at: https://docs.deno.com/runtime/manual/tools/unstable_flags
> 
> create-svelte version 6.0.10
> 
> T Welcome to SvelteKit! |
> 
>     * Which Svelte app template?
>       |  > SvelteKit demo app (A demo app showcasing some of the features of SvelteKit - play a word guessing game that works without
>       JavaScript!)
>       |    Skeleton project
>       |    Library project
>       —
>       error: Top-level await promise never resolved
>       at  (file:///C:/Users/lfmaciasr/AppData/Local/deno/npm/registry.npmjs.org/create-svelte/6.0.10/bin.js:45:17)
> 
> 
> Note: the create wizard worked before upgrading.

I got the same error and fixed by upgrading to the canary version. v1.42.1 is out now, and should fix your issue with https://github.com/denoland/deno/pull/23044

**Comment by lfmaciasr** [comment]
Thanks for the tip! After upgrading I get the same error as you when triying deno task build.

**Comment by littledivy** [comment]
![image](https://github.com/denoland/deno/assets/34997667/f97d094c-86cc-43bd-821f-981e975cb422)


SvelteKit Vite development server works with  `DENO_FUTURE=1 deno task dev`. 

There is one error blocking `deno task build` support: https://github.com/denoland/deno/issues/23455

**Comment by bachrc** [comment]
Oh, is it fixed? How can we manage to get it to work?

**Comment by flevi29** [comment]
I don't think it is released just yet. Even if it works, you still might need Node.js for linting rules, `svelte-check`, formatting .svelte files, separate tsconfig for the Svelte LSP and probably more, much of which is a separate issue from actually being able to run the thing with Deno.
Much of which I have no idea how they'll ever solve any other way than running ESLint, Prettier, TypeScript through Deno as well, essentially leaving any internal tooling untouched, thus most of the benefits of using Deno being nullified. I imagine this will be the case with any FE framework.
And there might be bugs.

But a milestone is a milestone, I still would like to not have to install both Deno and Node.js.

**Comment by birkskyum** [comment]
Fantastic, will give it a try with `deno upgrade --canary` when the CI is through

**Comment by birkskyum** [comment]
@littledivy  Things are much better, thanks! I just want to point out that when running i.e.:

deno run -A npm:create-svelte@latest my-sveltekit-app
deno task dev

Then going to the localhost address will error with a :

```sh
11:34:45 AM [vite] Error when evaluating SSR module /src/routes/+page.svelte: failed to import "/src/routes/Counter.svelte"
|- ReferenceError: requestAnimationFrame is not defined
    at eval (/Users/admin/repos/deno-kitchensink/my-sveltekit-app/node_modules/.deno/svelte@5.0.0-next.128/node_modules/svelte/src/internal/client/timing.js, <anonymous>:8:45)
    at async instantiateModule (file:///Users/admin/repos/deno-kitchensink/my-sveltekit-app/node_modules/.deno/vite@5.2.11/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

ReferenceError: requestAnimationFrame is not defined
    at eval (/Users/admin/repos/deno-kitchensink/my-sveltekit-app/node_modules/.deno/svelte@5.0.0-next.128/node_modules/svelte/src/internal/client/timing.js, <anonymous>:8:45)
    at async instantiateModule (file:///Users/admin/repos/deno-kitchensink/my-sveltekit-app/node_modules/.deno/vite@5.2.11/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9) {
  stack: "ReferenceError: requestAnimationFrame is not defined\n" +
    "    at eval (/Users/admin/repos/deno-kitchensin"... 311 more characters
}
```

But it works if instead using:

```sh
pnpm i
DENO_FUTURE=1 deno task dev
```

**Comment by littledivy** [comment]
It should work with `DENO_FUTURE=1` ...which will become the default in Deno 2.

```
deno run -A npm:create-svelte@latest

DENO_FUTURE=1 deno task dev
DENO_FUTURE=1 deno task build
```

**Comment by phoenisx** [comment]
Hi Team

Thanks for all the efforts. 🙇🏽 
I was trying out Deno with SvelteKit from scratch and thus tried using Svelte v5 with `create-svelte` scaffolding.
Most of the things work fine, but after building the app using `deno task build`, `deno task preview` does not load website on browser, the command runs fine though.

This doesn't happen with Svelte v4.

**Comment by phoenisx** [comment]
There's one more issue I think, not sure though.

Usually when building the SvelteKit app with `adapter-node` (I think that's the default when using with `deno`), the output is generated in `build` or whatever the `out` directory is set in the `svelte.config.js` `adapter` configuration.

Using `deno task build` doesn't generate the bundle inside (declared) `out` folder.

Am I missing something or are the generated files bundled in another folder?
I tried `deno task build` both with svelte v4 and v5, `deno task preview` works, but I wanted to start the server using `deno` cli and couldn't find the generated server script.

**Comment by littledivy** [comment]
@phoenisx Can you provide steps to reproduce this?

Svelte 5 works fine for me:

```
$ deno run -A npm:create-svelte@latest

create-svelte version 6.3.9

┌  Welcome to SvelteKit!
│
◇  Where should we create your project?
│    (hit Enter to use current directory)
│
◇  Which Svelte app template?
│  SvelteKit demo app
│
◇  Add type checking with TypeScript?
│  Yes, using TypeScript syntax
│
◇  Select additional options (use arrow keys/space bar)
│  Try the Svelte 5 preview (unstable!)
│
└  Your project is ready!

$ deno i
$ deno task build
$ deno task preview
Task preview vite preview
  ➜  Local:   http://localhost:4173/
  ➜  Network: use --host to expose
  ➜  press h + enter to show help
```

**Comment by phoenisx** [comment]
@littledivy I am using Skeleton project, which is not working for me

```
$ deno run -A npm:create-svelte@latest sveltekit-deno-v5

create-svelte version 6.3.9

┌  Welcome to SvelteKit!
│
◇  Which Svelte app template?
│  Skeleton project
│
◇  Add type checking with TypeScript?
│  Yes, using TypeScript syntax
│
◇  Select additional options (use arrow keys/space bar)
│  Try the Svelte 5 preview (unstable!)
│
└  Your project is ready!

$ deno task build
$ deno task preview
```

**Comment by littledivy** [comment]
Thanks. I can reproduce it!

**Comment by phoenisx** [comment]
There's one more issue I've found.
After generating the build, running the server using `deno` cli doesn't work, cause a lot of node imports in generated scripts are not using `node:` prefix.

I am not able to find any options in `deno run` that can force load such node imports. I remember there was some unstable flag to force import such node imports.

_**Can u point me how I can run the generated server script using `deno` cli?**_

**Comment by eltigerchino** [comment]
> ~There's one more issue I've found.~ ~After generating the build, running the server using `deno` cli doesn't work, cause a lot of node imports in generated scripts are not using `node:` prefix.~
> 
> ~I am not able to find any options in `deno run` that can force load such node imports. I remember there was some unstable flag to force import such node imports.~
> 
> ~_**Can u point me how I can run the generated server script using `deno` cli?**_~
> 
> I found the option: it's `--unstable-sloppy-imports`. The script still fails to run using `deno` cli, cause it contains `process.env`. Trying to google it now. 🙇🏽
> 
> ### Update:
> Found that after generating the build, using `deno` cli causes following errors:
> 
> * `node:` prefix not used for some node imports
> * `process.env` used without `node:process` import
> * `setImmediate` used without `node:timers` import
> 
> All of these errors are limited to `build/{env, handler, index}.js` scripts. Fixing them I was able to run the server using `deno` cli (without `--unstable-*` options).

Could you share the exact error logs?

**Comment by phoenisx** [comment]
```
$ deno task build
$ deno run -A build/index.js

error: Relative import path "http" not prefixed with / or ./ or ../
If you want to use a built-in Node module, add a "node:" prefix (ex. "node:http").
    at sveltekit-deno/build/index.js:3:18

error: Relative import path "path" not prefixed with / or ./ or ../
If you want to use a built-in Node module, add a "node:" prefix (ex. "node:path").
    at sveltekit-deno/build/handler.js:6:42

error: Relative import path "fs" not prefixed with / or ./ or ../
If you want to use a built-in Node module, add a "node:" prefix (ex. "node:fs").
    at sveltekit-deno/build/handler.js:4:21

error: Relative import path "querystring" not prefixed with / or ./ or ../
If you want to use a built-in Node module, add a "node:" prefix (ex. "node:querystring").
    at sveltekit-deno/build/index.js:4:21

---

error: Uncaught (in promise) ReferenceError: process is not defined
        return prefixed in process.env ? process.env[prefixed] : fallback;
                           ^
    at env (sveltekit-deno/build/env.js:35:21)
    at sveltekit-deno/build/handler.js:1149:16

---

Listening on 0.0.0.0:3000
ReferenceError: setImmediate is not defined
    at ServerImpl.Polka.attach (sveltekit-deno/build/index.js:164:31)
    at ServerImpl.emit (ext:deno_node/_events.mjs:387:28)
    at node:http:1416:16
    at new Promise (<anonymous>)
    at handler (node:http:1414:16)
    at ext:deno_http/00_serve.ts:369:24
    at ext:deno_http/00_serve.ts:569:29
    at eventLoopTick (ext:core/01_core.js:175:7)

```

**Comment by eltigerchino** [comment]
> ```
> $ deno task build
> $ deno run -A build/index.js
> 
> error: Relative import path "http" not prefixed with / or ./ or ../
> If you want to use a built-in Node module, add a "node:" prefix (ex. "node:http").
>     at sveltekit-deno/build/index.js:3:18
> 
> error: Relative import path "path" not prefixed with / or ./ or ../
> If you want to use a built-in Node module, add a "node:" prefix (ex. "node:path").
>     at sveltekit-deno/build/handler.js:6:42
> 
> error: Relative import path "fs" not prefixed with / or ./ or ../
> If you want to use a built-in Node module, add a "node:" prefix (ex. "node:fs").
>     at sveltekit-deno/build/handler.js:4:21
> 
> error: Relative import path "querystring" not prefixed with / or ./ or ../
> If you want to use a built-in Node module, add a "node:" prefix (ex. "node:querystring").
>     at sveltekit-deno/build/index.js:4:21
> 
> ---
> 
> error: Uncaught (in promise) ReferenceError: process is not defined
>         return prefixed in process.env ? process.env[prefixed] : fallback;
>                            ^
>     at env (sveltekit-deno/build/env.js:35:21)
>     at sveltekit-deno/build/handler.js:1149:16
> 
> ---
> 
> Listening on 0.0.0.0:3000
> ReferenceError: setImmediate is not defined
>     at ServerImpl.Polka.attach (sveltekit-deno/build/index.js:164:31)
>     at ServerImpl.emit (ext:deno_node/_events.mjs:387:28)
>     at node:http:1416:16
>     at new Promise (<anonymous>)
>     at handler (node:http:1414:16)
>     at ext:deno_http/00_serve.ts:369:24
>     at ext:deno_http/00_serve.ts:569:29
>     at eventLoopTick (ext:core/01_core.js:175:7)
> ```

Thanks! I've created a SvelteKit PR here https://github.com/sveltejs/kit/pull/12641 . The `node:` prefix and node globals usage issues stem from the `adapter-node` package and [Polka](https://github.com/lukeed/polka/blob/main/packages/polka/index.js). We'll need to update both.

**Comment by phoenisx** [comment]
Created this post build script: https://gist.github.com/phoenisx/fbd795775284a7299af9ee5409e772c6 for fixing the generated server scripts in the meantime.

**Comment by phoenisx** [comment]
Apart from the above-mentioned issues. I found that when running `deno run -A build/index.js`, server requests are not triggered at all, i.e. `load` fn and default form actions are not triggered when launching the app or submitting a form respectively (vite:dev works fine though)

I tried debugging the code, but I couldn't find the root cause for this yet.

**Comment by benmccann** [comment]
SvelteKit has been updated. I have PRs out to `polka` as well. In the meantime, you might be able to use a custom server to use something other than `polka`: https://kit.svelte.dev/docs/adapter-node#custom-server. But hopefully that won't be needed for long as Luke said he'd try to take a look at my PRs to `polka` later today

**Comment by birkskyum** [comment]
Thanks @benmccann . Just to make it easier to track in here:
- https://github.com/lukeed/polka/pull/212
- https://github.com/lukeed/polka/pull/214

**Comment by benmccann** [comment]
A new version of polka has been released, so everything should hopefully have the required imports and `node:` prefixes now. Please test again and let me know if we missed any.

Do note that I did not require the latest `polka` in `@sveltejs/adapter-node`, so you will have to update `polka` to `1.0.0-next.27` in your lockfile (or blast away your `node_modules` and lockfile if that's easier for anyone not familiar with the commands to update the lockfile).

I'm also curious if any Deno folks have feedback about https://github.com/sveltejs/kit/issues/12678. I wonder if that user might be mistaken because I would be surprised it wasn't reported as an issue until now. I also really don't want to change SvelteKit in that manner so hope it could be implemented on the Deno side if there is some issue we could track for it.

**Comment by cotyhamilton** [comment]
Polka is a dev dependency in sveltejs/adapter-node, there's no reference to it in the lockfile. I guess the package gets built using a static version of polka (1.0.0-next.25)

**Comment by bartlomieju** [comment]
> I'm also curious if any Deno folks have feedback about https://github.com/sveltejs/kit/issues/12678. I wonder if that user might be mistaken because I would be surprised it wasn't reported as an issue until now. I also really don't want to change SvelteKit in that manner so hope it could be implemented on the Deno side if there is some issue we could track for it.

@benmccann looks like the user sorted out the problem after all. Thanks for all the updates to SvelteKit!

**Comment by benmccann** [comment]
> Polka is a dev dependency in sveltejs/adapter-node, there's no reference to it in the lockfile. I guess the package gets built using a static version of polka (1.0.0-next.25)

Ah, thanks for the heads up. I've bumped it and released a new version of `@sveltejs/adapter-node`. Let me know if there's anything remaining to make it work on Deno

**Comment by cotyhamilton** [comment]
build and preview work ❤️ 

```bash
deno run -A npm:create-svelte@latest
deno task build # or deno run -A npm:vite build
deno task preview # or deno run -A npm:vite preview
```

Unfortunately there are still some non `node:` prefixed imports in the `build/handler.js` when using the node adapter so we can't run the server directly with deno (I looked at the node adapter source briefly but it's not obvious to me where these are coming from)

```bash
deno run -A build/index.js
error: Relative import path "path" not prefixed with / or ./ or ../
  hint: If you want to use a built-in Node module, add a "node:" prefix (ex. "node:path").
    at file:///Users/cotyhamilton/dev/my-app/build/handler.js:7:42
```

```js
// build/handler.js

import './shims.js';
import fs$1, { createReadStream } from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import * as fs from 'fs';
import { readdirSync, statSync } from 'fs';
import { resolve, join, normalize } from 'path';
import * as qs from 'node:querystring';
import { fileURLToPath } from 'node:url';
import { Readable } from 'node:stream';
import { Server } from './server/index.js';
import { manifest, prerendered, base } from './server/manifest.js';
import { env } from './env.js';

// ...
```

**Comment by birkskyum** [comment]
@cotyhamilton , isn't it just the `import { resolve, join, normalize } from 'path';` line?

**Comment by eltigerchino** [comment]
These are coming `sirv` which is imported in the `handler.js` file before being bundled. I wonder if we can get rollup to prefix the node built-ins instead.

https://github.com/sveltejs/kit/blob/85f3ec7408f9cf553ebbaaf58e1bb09b8ea2edc9/packages/adapter-node/src/handler.js#L5

https://github.com/lukeed/sirv/blob/50b1964b8a8342e14a711d47f793298c2a7aeeb7/packages/sirv/index.js#L1-L2

cc: @benmccann @lukeed

**Comment by Conduitry** [comment]
I've just opened https://github.com/sveltejs/kit/pull/12785 in the Rollup config for SvelteKit's Node adapter to prefix the imports with `node:` at bundle time, which should hopefully address this without needing changes to the underlying libraries.

**Comment by Conduitry** [comment]
Perhaps during the bundling of the user's application, SvelteKit should also replace other imports that are coming from the user's dependencies that are getting bundled? I'm not sure. I don't know whether that ought to be SvelteKit's responsibility, or whether it ought to just not introduce any unprefixed imports of its own.

**Comment by benmccann** [comment]
Merged and released as `@sveltejs/adapter-node` version `5.2.6`. Please test and let us know if there's anything else that's left to be done

**Comment by cotyhamilton** [comment]
I have an SSR app I tried this out on and all of it functions correctly now!

**Comment by Conduitry** [comment]
Interesting. What sorts of errors appear in the console?

For the record, `vite preview` isn't the same thing as running what's in `build/` - and, in fact, is independent of what adapter you have selected. `vite preview` runs the code in `.svelte-kit/` in the form before the adapter gets its hands on it.

**Comment by karimfromjordan** [comment]
I just opened a PR https://github.com/lukeed/sirv/pull/163

**Comment by cotyhamilton** [comment]
@Conduitry there are no error messages in the console. It was talked about briefly above in this thread, but not much info in here on it

<img width="1440" alt="Screenshot 2024-10-10 at 4 58 11 PM" src="https://github.com/user-attachments/assets/19fb562a-c87b-4d75-b67a-6bce269cc812">

**Comment by marvinhagemeister** [comment]
FYI: The "ERR_EMPTY_RESPONSE" or "can't open the page" error is tracked here https://github.com/denoland/deno/issues/26115

**Comment by lukeed** [comment]
sirv@3.0.0 was released to support this

**Comment by mjdNjhNJ** [comment]
Hi, I'm getting following warning when running SvelteKit project:
```
Task dev vite dev
▲ [WARNING] Cannot find base config file "./.svelte-kit/tsconfig.json" [tsconfig.json]

    tsconfig.json:2:12:
      2 │   "extends": "./.svelte-kit/tsconfig.json",
        ╵              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Here is my project structure:
<img width="301" alt="image" src="https://github.com/user-attachments/assets/1d24681a-ea4b-4aa6-b1db-f11322f4f4d6">

I'm on Svelte 5 btw.

**Comment by dominikg** [comment]
that warning occurs when the generated types in `.svelte-kit` are not present yet, eg the very first time you call vite dev. You can call `svelte-kit sync` to create it beforehand. see https://kit.svelte.dev/docs/cli#svelte-kit-sync

**Comment by jakubdonovan** [comment]
> that warning occurs when the generated types in `.svelte-kit` are not present yet, eg the very first time you call vite dev. You can call `svelte-kit sync` to create it beforehand. see https://kit.svelte.dev/docs/cli#svelte-kit-sync
> 
> edit: please ask these questions on svelte discord: https://svelte.dev/chat to not spam all the lovely people in this specific issue with unrelated things

I still seem to be getting issues running sveltekit on Deno. 
`deno run -A create-svelte testing`
![Screenshot 2024-10-15 at 01 11 40](https://github.com/user-attachments/assets/6e3d4aec-8f4e-4014-a1ed-8fa44adae7b8)

What do you suggest?

**Comment by benmccann** [comment]
The latest SvelteKit is now on `sirv` 3. Happy to help review and land any other code changes needed for Deno support.

**Comment by jakubdonovan** [comment]
> The latest SvelteKit is now on `sirv` 3. Happy to help review and land any other code changes needed for Deno support.

Unfortunately, updating to 2.7.1 did not resolve the issue with the types for me. 
![Screenshot 2024-10-15 at 06 48 16](https://github.com/user-attachments/assets/3ded75c4-4a29-4b51-a0b1-80b7827dca75)
![Screenshot 2024-10-15 at 06 48 37](https://github.com/user-attachments/assets/77834517-af4d-4aa9-bbbc-8e399cbc6c9c)

**Comment by sspilleman** [comment]
> FYI: The "ERR_EMPTY_RESPONSE" or "can't open the page" error is tracked here #26115

Is this being worked on? I have the feeling the solution is to force "every package" to comply with the "node:" or "npm:" specifiers because deno needs it... It simply doesn't feel like "deno can be used as a drop-in replacement for node/bun" as stated in the "deno 2 release video/notes/etc." ... I'm ok with having issues, no probs at all... I like deno a lot, switched completely from nodejs and never looked back I'm used to having issues with a lot of npm packages, been there, done that.... Just don't state node compat if it requires the node ecosystem to change their code...

Would it be a temp solution to "just" try assuming without specifiers it's either node: or npm: for the time being?

**Comment by cam-mcevenue** [comment]
Any idea if there is a way to get the Deno LSP to resolve sveltekit modules, generated types and path aliases? 

Can't resolve any of the `./$types` generated types for routes, or any of the `$app/*` types, or `$lib`

Not sure if I need to change something in my deno.json. the Compiler Options don't support rootDir or paths as of right now, so not sure if this is possible. I have to disable the LSP in favour of sveltes LSP to get types/intellisense to work

**Comment by cam-mcevenue** [comment]
Any idea how to get sveltekit generated types, modules and path aliases to work with deno LSP in VSCode? The svelte LSP resolves types fine in .svelte files. Deno LSP doesn't check those currently. But all the other .ts files like +page can't resolve `$app/*` types, `./$types` generated types, or `$lib` path alias.

deno.json typescript compiler options doesn't yet support `path` or `rootDir` config.

### To reproduce

Run `deno init` in an empty repo, clear out the deno.json config and files generated.

`deno run -A npm:create-svelte@latest` and install the sverdle demo project

`deno install --allow-scripts`

`deno run dev` to generate the types

Open any of the .ts files that use `$app/*`, `$lib` or `./$types`. You'll see the deno-ts errors I'm talking about

**Comment by jakubdonovan** [comment]
> Any idea how to get sveltekit generated types, modules and path aliases to work with deno LSP in VSCode? The svelte LSP resolves types fine in .svelte files. Deno LSP doesn't check those currently. But all the other .ts files like +page can't resolve `$app/*` types, `./$types` generated types, or `$lib` path alias.
> 
> deno.json typescript compiler options doesn't yet support `path` or `rootDir` config.
> 
> ### To reproduce
> Run `deno init` in an empty repo, clear out the deno.json config and files generated.
> 
> `deno run -A npm:create-svelte@latest` and install the sverdle demo project
> 
> `deno install --allow-scripts`
> 
> `deno run dev` to generate the types
> 
> Open any of the .ts files that use `$app/*`, `$lib` or `./$types`. You'll see the deno-ts errors I'm talking about
> 
> **Current workaround** is to disable the Deno LSP for the src directory
> 
> ```
> .vscode/settings.json
> 
> {
>   "deno.enable": true,
>   "deno.lint": true,
>   "deno.disablePaths": ["./src"]
> }
> ```

Yes, I would also like to know if there is a way to get denols to properly pick up sveltekit generated types.
![Screenshot 2024-10-17 at 07 51 30](https://github.com/user-attachments/assets/bdb96b27-6785-4b2a-beb5-0847e9f3b4b1)

**Comment by lts20050703** [comment]
@cam-mcevenue @jakubdonovan At least for $lib, it can be solved by adding a import map to `deno.json` or `deno.jsonc`

```jsonc
{
	// ...
	"imports": { "$lib": "./src/lib", "$lib/": "./src/lib/" }
	// ...
}
```
I looked at the SvelteKit tsconfig file (`./.svelte-kit/tsconfig.json`) and only saw mentions of `$lib` in `compilerOptions.paths`. @benmccann do you know what `$env` and `$app` (as well as `./$types`) maps to?

**Comment by jakubdonovan** [comment]
@benmccann Any update on this?

**Comment by StevenStavrakis** [comment]
> @cam-mcevenue @jakubdonovan At least for $lib, it can be solved by adding a import map to `deno.json` or `deno.jsonc`
> 
> ```js
> {
> 	// ...
> 	"imports": { "$lib": "./src/lib", "$lib/": "./src/lib/" }
> 	// ...
> }
> ```
> 
> I looked at the SvelteKit tsconfig file (`./.svelte-kit/tsconfig.json`) and only saw mentions of `$lib` in `compilerOptions.paths`. @benmccann do you know what `$env` and `$app` (as well as `./$types`) maps to?

I could not get the `deno.json` import paths to work for this. I ended up creating a `tsconfig` file and putting the import paths in there, and now it works fine, though I don't know what side effects that might have for the project as a whole, if any.

**Comment by Malix-Labs** [comment]
Also see https://github.com/sveltejs/cli/issues/214

**Comment by StevenStavrakis** [comment]
Interestingly, the solution provided in [the comment above](https://github.com/denoland/deno/issues/17248#issuecomment-2419397683) works for env variables, but not for `$lib`

**Comment by phoenisx** [comment]
Not sure if this works already, cause I wasn't able to figure out a proper solution to this.

I'm trying to create a barrel file to export specific Svelte Components from a folder, but Deno lint is throwing an error, also Deno is unable to resolve the import/export statements for `*.svelte` modules.

E.g. Following doesn't work for me, maybe I'm doing something wrong.
```ts
export { default as Component } from "./Component.svelte";
``` 

Maybe this works: `declare module "*.svelte" {}` in `global.d.ts` to resolve all svelte imports in `.ts` modules, but this might affect Svelte Component typings.

**Comment by Malix-Labs** [comment]
> Interestingly, the solution provided in [the comment above](https://github.com/denoland/deno/issues/17248#issuecomment-2419397683) works for env variables, but not for `$lib`

Did you found a fix @StevenStavrakis ?

**Comment by StevenStavrakis** [comment]
> > Interestingly, the solution provided in [the comment above](https://github.com/denoland/deno/issues/17248#issuecomment-2419397683) works for env variables, but not for `$lib`
> 
> Did you found a fix @StevenStavrakis ?

Only thing that worked for me was creating a tsconfig and doing the import aliasing there. Ultimately, I moved my project off of Deno and onto Bun, but I'm not using any special Bun features so it might as well be on standard node with pnpm at this point.

**Comment by Malix-Labs** [comment]
I have some rough edge with svelte on deno too

**Comment by jakubdonovan** [comment]
> > > Interestingly, the solution provided in [the comment above](https://github.com/denoland/deno/issues/17248#issuecomment-2419397683) works for env variables, but not for `$lib`
> > 
> > 
> > Did you found a fix @StevenStavrakis ?
> 
> Only thing that worked for me was creating a tsconfig and doing the import aliasing there. Ultimately, I moved my project off of Deno and onto Bun, but I'm not using any special Bun features so it might as well be on standard node with pnpm at this point.

Yep, that's what I did, after about two days of battling issue after issue and no real response from the Deno staff. I really like what they're doing, but unfortunately support is not all there yet. I will try again in 6 months and see how things have changed.

**Comment by bartlomieju** [comment]
Hey folks sorry for slow response, we got a ton of issues getting opened after the release of Deno 2 and we're trying to tackle them to unblock as many use cases as soon as possible.

We understand that you are all excited to use Svelte/SvelteKit with Deno and we have not forgotten about it. We're addressing stuff like https://github.com/denoland/deno/pull/26311 to make it easier to use Svelte with Deno. We also have a few bigger challenges like figuring out how to make auto-generated types work properly as well as how to make Svelte language server and Deno's language server play nicely.

I can assure you that we will tackle these things one, by one, so please bear with us. If you hit a specific problem it might be beneficial to help us prioritize if you open a separate issue - it's easy to lose something in this very long thread.

Thanks for your patience.

**Comment by Malix-Labs** [comment]
https://github.com/denoland/deno/issues/25342 has just been closed :tada:

**Comment by bartlomieju** [comment]
Hey again folks 🖐️ 

So I was about to write out separate GH issue for various problems we distilled from this issue in the past few weeks and... I have trouble doing so.

We identified several, distinct problems:
1. The main issue using Svelte is with the LSP, there's a lot of red squiggly lines in every `.svelte` file
2. Having `unstable: sloppy-imports` still requires to exclude `no-slippy-imports` lint rule
3. If there’s no `unstable-sloppy-import` imports enabled, then most imports in `.svelte` files are resulting in LSP `deno(no-local)` diagnostics. But that diagnostics doesn’t suggest a solution like adding `unstable-sloppy-imports`.
4. `./$types` imports are not resolved and result in the same error. They use `tsconfig.json` and use `rootDirs` option.
5. Another big problem is that we use “NodeNext”-ish resolution but it uses “bundler” resolution

That said, I wanted to test this from scratch, so I did:
```
deno run -A -r npm:create-svelte@latest
```
Which resulted in:
```
'npm create svelte' has been replaced with 'npx sv create'
```

Fine, I did:
```
deno run -A --reload npm:sv create svelte-kit-app
```
Used default options, as well as a lot of additional dependencies like `tailwind`, `playwright`, etc.

But when I opened the VSCode and installed Svelte extension (again, starting from scratch), all seems fine, with no errors and Intellisense working:

![Screenshot 2024-11-07 at 02 08 41](https://github.com/user-attachments/assets/04587689-b404-4ed0-a9bb-abe9fd5577c2)

---

May I kindly ask you to open separate issues with problems you are running into in your existing projects or new ones that you create? This will greatly help us identify common problems that need to be addressed one-by-one. At this point this issue is a catch-all and makes it hard to figure out what are the pitfalls you are running into :(

Thanks again!

**Comment by vladjerca** [comment]
> @jakubdonovan I was able to solve `$env` with the following additions to the import map
> 
> ```js
> // deno.json or deno.jsonc
> {
> 	"imports": {
>.         <REDACTED>
> 	}
> }
> ```
> 
> As for `./$types`, I found that it maps to `./svelte-kit/types/src/routes/.../$types.d.ts`.
> 
> For example, in the file `./src/routes/api/version/+server.ts`, importing `./$types` will map to `./.svelte-kit/types/src/routes/api/version/$types.d.ts`
> 
> Of course it is possible to just list out all of the paths of the SvelteKit app, but I think you can imagine how tedious it would be.
> 
> @bartlomieju I wonder if the import map support like somehow reading the file path and then substitute it? I imagine something like `"./src/routes/**/$types": "./.svelte-kit/types/src/routes/**/$types.d.ts"`
> 
> Finally, for $app stuff, it looks like it maps to `@sveltejs/kit/src/runtime/app/...`, for example `$app/environment` maps to `@sveltejs/kit/src/runtime/app/environment/index.js`. However, because the package.json `exports` field doesn't export that path, it is impossible to import it. I had to add something like the following
> 
> ```js
> // @sveltejs/kit/package.json
> {
> 	"exports": {
> 		// ...
> 		"./environment": {
> 			"types": "./src/runtime/app/environment/types.d.ts",
> 			"import": "./src/runtime/app/environment/index.js"
> 		}
> 	}
> }
> ```
> 
> and then add the following to the import path
> 
> ```js
> 	"imports": {
> 		// ...
> 		"$app/environment": "npm:@sveltejs/kit/environment",
> 	}
> ```
> 
> for it to work. @benmccann Is there any reason you guys don't export the src/runtime/app/ stuff?

For those brave souls seeking to harness the untamed power of Deno LSP, here lies the forbidden configuration.

```json
{
  "imports": {
    "$lib": "./src/lib",
    "$lib/": "./src/lib/",
    "$env/static/public": "./.svelte-kit/ambient.d.ts",
    "$env/static/private": "./.svelte-kit/ambient.d.ts",
    "$env/dynamic/public": "./.svelte-kit/ambient.d.ts",
    "$env/dynamic/private": "./.svelte-kit/ambient.d.ts",
    "$app/forms": "../../node_modules/@sveltejs/kit/types/index.d.ts",
    "$app/environment": "../../node_modules/@sveltejs/kit/types/index.d.ts",
    "$app/navigation": "../../node_modules/@sveltejs/kit/types/index.d.ts",
    "$app/server": "../../node_modules/@sveltejs/kit/types/index.d.ts",
    "$app/stores": "../../node_modules/@sveltejs/kit/types/index.d.ts"
  }
}
```

But **heed this warning**, intrepid coder: The `node_modules` path is a fickle beast. Adapt it to your workspace, lest it consume your project whole.

**Comment by vladjerca** [comment]
@bartlomieju is there any timeline for supporting the `./$types` imports generated by SvelteKit. It's a bit unclear when to expect full support (even knowing that it would take 6 months is better than limbo 😅).

I appreciate all the work you are guys doing but if there's any plan some info would help us also make decisions for our projects. 

Personally I would like to avoid going down the `"deno.disablePaths": ["./src"]` route 👀

**Comment by bartlomieju** [comment]
Hey folks, Deno 2.1.8 should have a much better support for using Svelte/SvelteKit projects, with these recently landed PRs:
- https://github.com/denoland/deno/pull/27855
- https://github.com/denoland/deno/pull/27690
- https://github.com/denoland/deno/pull/27686
- https://github.com/denoland/deno/pull/27631
- and more...
Deno should properly resolve `./$types` and Vite virtual modules.

**Comment by AndreasNasman** [comment]
@bartlomieju Do you need some specific config for `./$types` resolution to work correctly? I receive the same error messages on v2.1.9 as @vladjerca when creating a "Hello World" project from this template that I made: https://github.com/AndreasNasman/deno-sveltekit-template/wiki

**Comment by bartlomieju** [comment]
That seems like a bug that @nayeemrmn is working on a fix for right now

**Comment by bartlomieju** [comment]
With the `./$types` issue fixed in Deno v2.1.10 we feel the phase 1 of Svelte/SvelteKit support is done.

There might still be some rough edges in the SvelteKit support - we'd love to fix them and continue to appreciate the reports you provide. But we've outgrown this issue - it's over 2 years old. At this point it will be easier to track other problems in dedicated issues.

Thanks everyone for chiming in on this issue and helping to improve Deno support for Svelte!
